ably/realtime_conn.go

```diff
@@ -950,10 +969,15 @@ func (c *Connection) lockSetState(state ConnectionState, err error, retryIn time
 }
 
 func (c *Connection) ctxCancelOnStateTransition() (context.Context, context.CancelFunc) {
+	// Returns parent context for retry process
 	ctx, cancel := context.WithCancel(context.Background())
 
-	off := c.internalEmitter.OnceAll(func(ConnectionStateChange) {
-		cancel()
+	off := c.internalEmitter.OnceAll(func(change ConnectionStateChange) {
+		// RTN12d - cancel the retry when connectionStateClosed
+		if change.Current == ConnectionStateClosed {
```

Comment:
I'll copy the relevant part from my previous comment:

> The purpose of this function is to stop the timer that moves the connection state from DISCONNECTED or SUSPENDED to CONNECTING. This should happen when:
> 
> * `connect` is called ([RTN11c](https://docs.ably.io/client-lib-development-guide/features/#RTN11c)).
> * `close` is called ([RTN12d](https://docs.ably.io/client-lib-development-guide/features/#RTN12d)).
> 
> This isn't captured by this change; `connect` causes a transition to CONNECTING, and `close` causes a transition to CLOSING. Maybe those were the states you meant to put here?

Reply:
> Currently, we are not using `retry` value anywhere, it's just a placeholder.

Not true, it's used in the `RetryIn` property of the `ConnectionEvent`: https://github.com/ably/ably-go/blob/083f8e334bc0cb55791f88223ed1c4e99a39b5ec/ably/realtime_conn.go#L912

> I find this solution more feasible because in other cases, things just makes code more messy

Sorry, I didn't make my suggestions clear enough because they actually simplify all this logic quite a bit. For instance:

> That way, it's easy to detect from where `connecting` is being called.

This just isn't necessary. The only reason it's necessary as things stand now is because of poorly factored code.

So, given I didn't manage to explain myself with words, here's my proposal in code: https://github.com/ably/ably-go/compare/fix/262-connection-close...fix/262-connection-close-alt As I hope you'll agree with, the overall logic is way simpler now, with less things interacting in non-obvious ways, with more explanatory naming, and more comments.

I'd really strongly prefer that to be the way forward. If you don't agree, I'll need to ask for a third opinion before reaching a decision.

> and we might need to go away from spec

This may be so, depending on the interpretation on what "aborts the retry process" means. I believe https://github.com/ably/docs/pull/1090 is a better way of defining what RTN11c is _actually_ trying to accomplish.

Reply:
No. it's just writing the value to `RetryIn`, which is literally not being used anywhere. So, technically, it means, it's not being used. I checked at all places, we just write the value and not read it for doing any processing. In the new code, we will use 
it for cancelling the loop

Reply:
I need to properly review your code, if it works, I am fine with it.

Reply:
> No. it's just writing the value to `RetryIn`, which is literally not being used anywhere. So, technically, it means, it's not being used. I checked at all places, we just write the value and not read it for doing any processing.

It's part of the public API. We're not the ones using it, clients are.

Reply:
Also, in the alt code, I think we might need to tweak it a bit, since it doesnt cover cancelling the loop when closed is called. There is explicit check to return nil error, but before that we set state to `connecting`, so it doesn't really cover the case. I am waiting for someone to accept the spec, if new spec changes are considered, I am happy to merge your code. Although, its not straightforward and combines disconnected & suspended spec in one loop. I would generally consider `WET ( write everything twice)` principal in the case, because it makes code easy to understand.

Reply:
> Also, in the alt code, I think we might need to tweak it a bit, since it doesnt cover cancelling the loop when closed is called. There is explicit check to return nil error, but before that we set state to `connecting`, so it doesn't really cover the case.

Oops, forgot about that. Addressed at bd8c500.



> I am waiting for someone to accept the spec, if new spec changes are considered, I am happy to merge your code.

@ably/client-library-sdk-admins Maybe we should merge https://github.com/ably/docs/pull/1090 it already? It has several approvals.



> Although, its not straightforward and combines disconnected & suspended spec in one loop. I would generally consider `WET ( write everything twice)` principal in the case, because it makes code easy to understand.

I decided to do it in one loop because the only change between DISCONNECTED and SUSPENDED is the waiting time between reconnection attempts, and that can be just a variable. In contrast, the logic around timer cancellation is a bit more complex and is the same, so I'd rather not write it twice.

Reply:
> No it's not being used clients. I checked, clients just write value to retryIn and not read from it

I meant end-users. They can access the field when they get a ConnectionEvent.


> > I am waiting for someone to accept the spec, if new spec changes are considered, I am happy to merge your code.
> 
> @ably/client-library-sdk-admins Maybe we should merge [ably/docs#1090](https://github.com/ably/docs/pull/1090) it already? It has several approvals.

I just realized it's been merged for a few hours now.